edgecenter/utils.go

```diff
@@ -734,3 +737,215 @@ func extractListenerIntoMap(listener *listeners.Listener) map[string]interface{}
+	if err != nil {
+		return err
+	}
+
+	err = tasks.WaitTaskAndProcessResult(client, results.Tasks[0], true, InstanceCreatingTimeout, func(task tasks.TaskID) error {
+		if taskInfo, err := tasks.Get(client, string(task)).Extract(); err != nil {
+			return fmt.Errorf("cannot get task with ID: %s. Error: %w, task: %+v", task, err, taskInfo)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func attachInterfaceToInstance(client *edgecloud.ServiceClient, instanceID string, iface map[string]interface{}) error {
+	iType := types.InterfaceType(iface["type"].(string))
+	opts := instances.InterfaceInstanceCreateOpts{
+		InterfaceOpts: instances.InterfaceOpts{Type: iType},
+	}
+
+	switch iType { //nolint: exhaustive
+	case types.SubnetInterfaceType:
+		opts.SubnetID = iface["subnet_id"].(string)
+	case types.AnySubnetInterfaceType:
+		opts.NetworkID = iface["network_id"].(string)
+	case types.ReservedFixedIpType:
+		opts.PortID = iface["port_id"].(string)
+	}
+	opts.SecurityGroups = getSecurityGroupsIDs(iface["security_groups"].([]interface{}))
+
+	log.Printf("[DEBUG] attach interface: %+v", opts)
+	results, err := instances.AttachInterface(client, instanceID, opts).Extract()
+	if err != nil {
+		return fmt.Errorf("cannot attach interface: %s. Error: %w", iType, err)
+	}
+
+	err = tasks.WaitTaskAndProcessResult(client, results.Tasks[0], true, InstanceCreatingTimeout, func(task tasks.TaskID) error {
+		taskInfo, err := tasks.Get(client, string(task)).Extract()
+		if err != nil {
+			return fmt.Errorf("cannot get task with ID: %s. Error: %w, task: %+v", task, err, taskInfo)
+		}
+
+		if _, err := instances.ExtractInstancePortIDFromTask(taskInfo); err != nil {
+			reservedFixedIPID, ok := (*taskInfo.Data)["reserved_fixed_ip_id"]
+			if !ok || reservedFixedIPID.(string) == "" {
+				return fmt.Errorf("cannot retrieve instance port ID from task info: %w", err)
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func removeSecurityGroupFromInstance(clientSG, clientInstance *edgecloud.ServiceClient, instanceID, portID string, removeSGs []edgecloud.ItemID) error {
+	for _, sg := range removeSGs {
+		sgInfo, err := securitygroups.Get(clientSG, sg.ID).Extract()
+		if err != nil {
+			return err
+		}
+
+		portSGNames := instances.PortSecurityGroupNames{PortID: &portID, SecurityGroupNames: []string{sgInfo.Name}}
+		sgOpts := instances.SecurityGroupOpts{PortsSecurityGroupNames: []instances.PortSecurityGroupNames{portSGNames}}
+
+		log.Printf("[DEBUG] remove security group opts: %+v", sgOpts)
+		if err := instances.UnAssignSecurityGroup(clientInstance, instanceID, sgOpts).Err; err != nil {
+			return fmt.Errorf("cannot remove security group. Error: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func attachSecurityGroupToInstance(clientSG, clientInstance *edgecloud.ServiceClient, instanceID, portID string, addSGs []edgecloud.ItemID) error {
```

Comment:
The correct name is `sgClient` and `instanceClient` 

Now it's confusing and can be read as "Security group of a client", not a "client for requesting SGs"